### PR TITLE
[AIROCMLIR-508] Attention masking + softmax calculation improvements

### PR DESCRIPTION
**Pull request by justinrosner**
## Motivation

This PR improves on some of the conditional logic used for masking in the main attention loop, as well as some fast math related changes. 

This implements: https://amd-hub.atlassian.net/browse/AIROCMLIR-508

## Technical Details

1. **Stop expanding float min/max ops**
    * Introduced a new `includeFloatMinMax` option. When set to false `arith.maxnumf`, `arith.minnumf`, `arith.maximumf`, `arith.minimumf` are not expanded into compare and select patterns (each single op was turned into 4 instructions). The AMDGPU backend natively supports these ops as single instructions and can pattern match them accordingly in iSel.
2. **`MaximumFOp` -> `MaxNumFOp` in softmax computation**
    * All four max-reduction operations (at least the float ones) were updated to `MaxNumFOp`. `MaximumFOp` propagates Nan, while `MaxNumFOp` does not. With `MaximumFOp` a NaN value immediately poisons the max, which makes every element's `exp(score - max)` produce Nan. With `MaxNumFOp` only the originally NaN elements in the score produce NaN in the exp result. However, the final output is going to be identical because the sum reduction over the exp results will always include a NaN.
3. **Use fast math flags on softmax division**
    * The final softmax normalization `output / sum_of_exp` was changed from a regular `arith::DivFOp` to one with `arith::FastMathFlags::arcp`. The divisor is always going to be >= 1.0 in valid cases since it accumulates exp2(score - max) values, and at least one score equals the max, contributing exp2(0) = 1. In the fully-masked case where all scores are -inf the sum accumulates to NaN, and the output of rcp(NaN) = NaN.
4. **Improve when causal masking is applied**
    * Causal masking masks positions where key_pos > query_pos (i.e., a token can't attend to future tokens). The attention M-loop iterates over blocks of K (key) positions. For a given Q block, early M-loop iterations cover low K positions that are all ≤ the Q positions in the block, so no masking is needed. Masking is only needed once the K block reaches the "diagonal" of the causal triangle, where K positions start exceeding Q positions.

## Test Plan

- PR CI
- Nightly CI

## Test Result

- [x] PR CI
- [x] Nightly CI

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

**Comment by justinrosner** [comment]
Note: I don't love the idea of making changes to external LLVM without upstreaming them, so does anyone have any ideas for how we can get around doing this?

File: external/llvm-project/mlir/include/mlir/Dialect/Arith/Transforms/Passes.td [hunk 1]
```diff
@@ -21,6 +21,11 @@ def ArithExpandOpsPass : Pass<"arith-expand"> {
               "Enable the F8E8M0 expansion patterns">,
        Option<"includeF4E2M1", "include-f4e2m1", "bool", /*default=*/"false",
               "Enable the F4E2M1 expansion patterns">,
+       Option<"includeFloatMinMax", "include-float-min-max", "bool",
+              /*default=*/"true",
+              "Expand arith.maximumf/minimumf/maxnumf/minnumf into "
+              "comparison+select patterns. Set to false when the target "
+              "natively supports these operations (e.g. AMDGPU backend).">,
```
Comment (dhernandez0) [inline external/llvm-project/mlir/include/mlir/Dialect/Arith/Transforms/Passes.td:28]: isn't there another way to do this? we can't (or at least don't want to) patch upstream in rocmlirTriton. Could we have a pass to do this instead?

Comment (dhernandez0) [reply external/llvm-project/mlir/include/mlir/Dialect/Arith/Transforms/Passes.td:28]: wait, I think it's fine because we don't use createArithExpandOpsPass in rocmlirTriton?

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 2]
```diff
@@ -957,9 +957,10 @@ struct BlockwiseReduceRewritePattern
         // Op verifier gurantees this.
         assert(rMethod == ReduceMethod::Max);
         if (elementType.isIntOrIndex()) {
-          kind = vector::CombiningKind::MAXIMUMF;
+          kind = vector::CombiningKind::MAXSI;
         } else {
-          kind = vector::CombiningKind::MAXIMUMF;
+
+          kind = vector::CombiningKind::MAXNUMF;
         }
```
[the remark this answers is not in the document]
Comment (dhernandez0) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:965]: @justinrosner

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 3]
```diff
@@ -979,7 +980,7 @@ struct BlockwiseReduceRewritePattern
       if (elementType.isIntOrIndex()) {
         reduced = arith::MaxSIOp::create(builder, loc, acc, input);
       } else {
-        reduced = arith::MaximumFOp::create(builder, loc, acc, input);
+        reduced = arith::MaxNumFOp::create(builder, loc, acc, input);
```
Comment (dhernandez0) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:986]: add a comment here explaining why we use MaxNumFOp

File: mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp [hunk 4]
```diff
@@ -865,7 +865,7 @@ struct GridwiseAttentionAccelRewritePattern
       Value ldgemm0OutBufferMax =
           InBoundsLoadOp::create(rewriter, loc, maxRowBufferElemType,
                                  gemm0OutBufferMax, gemm0OutBufferMaxCoords);
-      Value maxRowBufferNew = arith::MaximumFOp::create(
+      Value maxRowBufferNew = arith::MaxNumFOp::create(
```
Comment (dhernandez0) [inline mlir/lib/Dialect/Rock/Transforms/GridwiseGemmToBlockwise.cpp:869]: same comment here

**Comment by dhernandez0** [comment]
can you add benchmark numbers of before/after?

**Review by dhernandez0 — APPROVED** [review]

**Comment by dhernandez0** [comment]
> Note: I don't love the idea of making changes to external LLVM without upstreaming them, so does anyone have any ideas for how we can get around doing this?

we could have a pass that undoes the max expansion? that's ugly as well. Can we move that pass to happen later? when does the max -> v_max_f32 happen? We could look into how IREE/Triton do this.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp [hunk 5]
```diff
@@ -957,9 +957,10 @@ struct BlockwiseReduceRewritePattern
         // Op verifier gurantees this.
         assert(rMethod == ReduceMethod::Max);
         if (elementType.isIntOrIndex()) {
-          kind = vector::CombiningKind::MAXIMUMF;
```
Comment (umangyadav) [inline mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:960]: was this a bug ?

File: external/llvm-project/mlir/lib/Dialect/Arith/Transforms/ExpandOps.cpp [hunk 6]
```diff
@@ -744,14 +744,21 @@ struct ArithExpandOpsPass
...
-      arith::MinimumFOp,
-      arith::MaxNumFOp,
-      arith::MinNumFOp,
       arith::ScalingExtFOp,
       arith::ScalingTruncFOp
     >();
 
+    // Float min/max expansion is optional. Targets like AMDGPU natively
+    // support these via single instructions (e.g. v_max_f32, v_min_f32).
+    if (includeFloatMinMax) {
```
Comment (umangyadav) [inline external/llvm-project/mlir/lib/Dialect/Arith/Transforms/ExpandOps.cpp:753]: Need to add some tests for for this. 
Can we upstream these changes ?

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 5 above)
Comment (justinrosner) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:960]: Yeah, it looks like it was calling a float op for the int path

**Comment by justinrosner** [comment]
> > Note: I don't love the idea of making changes to external LLVM without upstreaming them, so does anyone have any ideas for how we can get around doing this?
> 
> we could have a pass that undoes the max expansion? that's ugly as well. Can we move that pass to happen later? when does the max -> v_max_f32 happen? We could look into how IREE/Triton do this.

It runs pretty early in the `buildBackendPipeline`, but the conversion to `v_max_f32` doesn't happen until much later. I did some investigation into what Triton does and it never has this problem because it just avoids running that pass all together. If I recall from my earlier investigation, I think IREE does the same thing.

File: external/llvm-project/mlir/lib/Dialect/Arith/Transforms/ExpandOps.cpp (hunk 6 above)
Comment (justinrosner) [reply external/llvm-project/mlir/lib/Dialect/Arith/Transforms/ExpandOps.cpp:753]: I added the tests. I'll open up a PR for this on upstream and see if they would be willing to accept it.

File: mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp (hunk 5 above)
Comment (umangyadav) [reply mlir/lib/Dialect/Rock/Transforms/BlockwiseGemmToThreadwise.cpp:960]: Can we add a test to catch this ?

**Review by umangyadav — APPROVED** [review]

**Comment by justinrosner** [comment]
> can you add benchmark numbers of before/after?

@dhernandez0 average improvement is +6% with these changes. The range is -1.82% to +18.58%. I've posted the full table in the linked ticket: https://amd-hub.atlassian.net/browse/AIROCMLIR-508
